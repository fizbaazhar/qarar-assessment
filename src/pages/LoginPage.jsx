import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, Link } from 'react-router-dom';
import { login, setError, setLoading } from '../redux/authSlice';
import AuthLayout from '../components/AuthLayout';
import Input from '../components/Input';
import Button from '../components/Button';

/**
 * LoginPage Component
 * Handles user authentication with email and password
 * Features: Form validation, error handling, and navigation to signup
 */
const LoginPage = () => {
  const [form, setForm] = useState({ email: '', password: '' });
  const [errors, setErrors] = useState({});
  
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const authError = useSelector((state) => state.auth.error);
  const isLoading = useSelector((state) => state.auth.isLoading);

  /**
   * Validates the login form
   * @returns {boolean} True if form is valid, false otherwise
   */
  const validateForm = () => {
    const newErrors = {};

    // Email validation
    if (!form.email.trim()) {
      newErrors.email = 'Email is required';
    } else if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(form.email)) {
      newErrors.email = 'Please enter a valid email address';
    }

    // Password validation
    if (!form.password) {
      newErrors.password = 'Password is required';
    } else if (form.password.length < 6) {
      newErrors.password = 'Password must be at least 6 characters';
    }

    setErrors(newErrors);
    return Object.keys(newErrors).length === 0;
  };

  const handleChange = (e) => {
    const { name, value } = e.target;
    setForm({ ...form, [name]: value });
    
    // Clear specific field error when user starts typing
    if (errors[name]) {
      setErrors({ ...errors, [name]: '' });
    }
    
    // Clear auth error when user starts typing
    if (authError) dispatch(setError(null));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    
    // Validate form before submission
    if (!validateForm()) {
      return;
    }

    dispatch(setLoading(true));
    
    try {
      // Simulate API call delay
      await new Promise(resolve => setTimeout(resolve, 1000));
      
      dispatch(login({ email: form.email, name: form.email.split('@')[0] }));
      navigate('/dashboard');
    } catch (err) {
      dispatch(setError('Login failed. Please try again.'));
    }
  };

  return (
    <AuthLayout 
      icon="ðŸ”’" 
      subtitle="Sign in to your account"
    >
      {/* Login Form */}
      <form onSubmit={handleSubmit} className="w-full space-y-6">
        <h2 className="text-2xl font-bold text-center text-gray-800 mb-6">
          Welcome Back
        </h2>
        
        {/* Email Field */}
        <Input
          label="Email Address"
          id="email"
          name="email"
          type="email"
          placeholder="Enter your email"
          value={form.email}
          onChange={handleChange}
          error={errors.email}
          disabled={isLoading}
        />

        {/* Password Field */}
        <Input
          label="Password"
          id="password"
          name="password"
          type="password"
          placeholder="Enter your password"
          value={form.password}
          onChange={handleChange}
          error={errors.password}
          disabled={isLoading}
        />

        {/* Auth Error Display */}
        {authError && (
          <div className="bg-red-50 border border-red-200 text-red-700 px-4 py-3 rounded-lg text-sm">
            {authError}
          </div>
        )}

        {/* Submit Button */}
        <Button
          type="submit"
          variant="primary"
          size="lg"
          loading={isLoading}
          className="w-full"
        >
          Sign In
        </Button>
      </form>

      {/* Sign Up Link */}
      <div className="mt-6 text-center">
        <p className="text-gray-600 text-sm">
          Don't have an account?{' '}
          <Link 
            to="/signup" 
            className="text-primary hover:text-primary-dark font-semibold hover:underline transition-colors"
          >
            Sign up here
          </Link>
        </p>
      </div>
    </AuthLayout>
  );
};

export default LoginPage; 