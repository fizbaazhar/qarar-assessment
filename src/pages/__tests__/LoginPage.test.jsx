import React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { vi } from 'vitest'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { configureStore } from '@reduxjs/toolkit'
import LoginPage from '../LoginPage'
import authReducer from '../../redux/authSlice'
import profileReducer from '../../redux/profileSlice'
import tasksReducer from '../../redux/tasksSlice'
import notificationsReducer from '../../redux/notificationsSlice'

// Mock react-router-dom
const mockNavigate = vi.fn()
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom')
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  }
})

// Mock the toaster utility
vi.mock('../../utils/toaster', () => ({
  showToast: vi.fn(),
}))

describe('LoginPage Component', () => {
  const user = userEvent.setup()
  
  // Create a test store
  const createTestStore = (preloadedState = {}) => {
    return configureStore({
      reducer: {
        auth: authReducer,
        profile: profileReducer,
        tasks: tasksReducer,
        notifications: notificationsReducer,
      },
      preloadedState,
    })
  }

  const renderWithProviders = (ui, { preloadedState = {}, store = createTestStore(preloadedState) } = {}) => {
    return render(
      <Provider store={store}>
        <BrowserRouter>
          {ui}
        </BrowserRouter>
      </Provider>
    )
  }

  beforeEach(() => {
    vi.clearAllMocks()
  })

  // Test basic rendering
  it('renders login form with all required elements', () => {
    renderWithProviders(<LoginPage />)
    
    expect(screen.getByText('Welcome Back')).toBeInTheDocument()
    expect(screen.getByLabelText('Email Address')).toBeInTheDocument()
    expect(screen.getByLabelText('Password')).toBeInTheDocument()
    expect(screen.getByRole('button', { name: 'Sign In' })).toBeInTheDocument()
    expect(screen.getByText("Don't have an account?")).toBeInTheDocument()
  })

  // Test form validation
  it('shows error when email is empty', async () => {
    renderWithProviders(<LoginPage />)
    
    const submitButton = screen.getByRole('button', { name: 'Sign In' })
    await user.click(submitButton)
    
    expect(screen.getByText('Email is required')).toBeInTheDocument()
  })

  it('shows error when password is empty', async () => {
    renderWithProviders(<LoginPage />)
    
    const emailInput = screen.getByLabelText('Email Address')
    await user.type(emailInput, 'test@example.com')
    
    const submitButton = screen.getByRole('button', { name: 'Sign In' })
    await user.click(submitButton)
    
    expect(screen.getByText('Password is required')).toBeInTheDocument()
  })

  it('prevents form submission with invalid email format', async () => {
    renderWithProviders(<LoginPage />)
    const emailInput = screen.getByLabelText('Email Address')
    await user.type(emailInput, 'justtext')
    const passwordInput = screen.getByLabelText('Password')
    await user.type(passwordInput, 'password123')
    
    const submitButton = screen.getByRole('button', { name: 'Sign In' })
    await user.click(submitButton)
    
    // Wait a bit for any async operations to complete
    await new Promise(resolve => setTimeout(resolve, 100))
    
    // Check that navigation did not occur (form validation prevented submission)
    expect(mockNavigate).not.toHaveBeenCalled()
    
    // Verify the form is still visible (not redirected)
    expect(screen.getByText('Welcome Back')).toBeInTheDocument()
  })

  it('shows error when password is too short', async () => {
    renderWithProviders(<LoginPage />)
    
    const emailInput = screen.getByLabelText('Email Address')
    await user.type(emailInput, 'test@example.com')
    
    const passwordInput = screen.getByLabelText('Password')
    await user.type(passwordInput, '123')
    
    const submitButton = screen.getByRole('button', { name: 'Sign In' })
    await user.click(submitButton)
    
    expect(screen.getByText('Password must be at least 6 characters')).toBeInTheDocument()
  })

  // Test successful form submission
  it('submits form successfully with valid data', async () => {
    renderWithProviders(<LoginPage />)
    
    const emailInput = screen.getByLabelText('Email Address')
    await user.type(emailInput, 'test@example.com')
    
    const passwordInput = screen.getByLabelText('Password')
    await user.type(passwordInput, 'password123')
    
    const submitButton = screen.getByRole('button', { name: 'Sign In' })
    await user.click(submitButton)
    
    // Wait for the async operations to complete
    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith('/dashboard')
    })
  })

  // Test form field interactions
  it('clears email error when user starts typing', async () => {
    renderWithProviders(<LoginPage />)
    
    const submitButton = screen.getByRole('button', { name: 'Sign In' })
    await user.click(submitButton)
    
    expect(screen.getByText('Email is required')).toBeInTheDocument()
    
    const emailInput = screen.getByLabelText('Email Address')
    await user.type(emailInput, 't')
    
    expect(screen.queryByText('Email is required')).not.toBeInTheDocument()
  })

  it('clears password error when user starts typing', async () => {
    renderWithProviders(<LoginPage />)
    
    const emailInput = screen.getByLabelText('Email Address')
    await user.type(emailInput, 'test@example.com')
    
    const submitButton = screen.getByRole('button', { name: 'Sign In' })
    await user.click(submitButton)
    
    expect(screen.getByText('Password is required')).toBeInTheDocument()
    
    const passwordInput = screen.getByLabelText('Password')
    await user.type(passwordInput, 'p')
    
    expect(screen.queryByText('Password is required')).not.toBeInTheDocument()
  })

  // Test navigation link
  it('has working signup link', () => {
    renderWithProviders(<LoginPage />)
    
    const signupLink = screen.getByText('Sign up here')
    expect(signupLink).toHaveAttribute('href', '/signup')
  })

  // Test accessibility
  it('has proper form labels and associations', () => {
    renderWithProviders(<LoginPage />)
    
    const emailInput = screen.getByLabelText('Email Address')
    const passwordInput = screen.getByLabelText('Password')
    
    expect(emailInput).toHaveAttribute('type', 'email')
    expect(passwordInput).toHaveAttribute('type', 'password')
  })

  it('has proper ARIA attributes', () => {
    renderWithProviders(<LoginPage />)
    
    const emailInput = screen.getByLabelText('Email Address')
    const passwordInput = screen.getByLabelText('Password')
    
    expect(emailInput).toHaveAttribute('id', 'email')
    expect(passwordInput).toHaveAttribute('id', 'password')
  })

  // Test error handling
  it('displays auth error when login fails', async () => {
    // Mock a failed login by setting up the store with an error
    const preloadedState = {
      auth: {
        error: 'Login failed. Please try again.',
        isLoading: false,
        isAuthenticated: false,
        user: null
      }
    }
    
    renderWithProviders(<LoginPage />, { preloadedState })
    
    expect(screen.getByText('Login failed. Please try again.')).toBeInTheDocument()
  })

  // Test form reset - this test needs to be updated since the form doesn't automatically reset
  it('maintains form state during component lifecycle', async () => {
    const { rerender } = renderWithProviders(<LoginPage />)
    
    const emailInput = screen.getByLabelText('Email Address')
    const passwordInput = screen.getByLabelText('Password')
    
    await user.type(emailInput, 'test@example.com')
    await user.type(passwordInput, 'password123')
    
    expect(emailInput).toHaveValue('test@example.com')
    expect(passwordInput).toHaveValue('password123')
    
    // Re-render the component
    rerender(
      <Provider store={createTestStore()}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider>
    )
    
    // The form should maintain its state since it's not explicitly reset
    expect(emailInput).toHaveValue('test@example.com')
    expect(passwordInput).toHaveValue('password123')
  })

  // Test snapshot
  it('matches snapshot', () => {
    const { container } = renderWithProviders(<LoginPage />)
    expect(container).toMatchSnapshot()
  })
}) 