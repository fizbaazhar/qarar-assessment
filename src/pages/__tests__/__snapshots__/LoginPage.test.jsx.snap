// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage Component > matches snapshot 1`] = `
<div>
  <div
    class="min-h-screen flex flex-col justify-center items-center bg-gradient-to-br from-primary/10 via-white to-primary/20"
  >
    <div
      class="w-full max-w-md p-8 bg-white rounded-2xl shadow-2xl flex flex-col items-center"
    >
      <div
        class="mb-6 flex flex-col items-center"
      >
        <div
          class="w-16 h-16 rounded-full bg-primary flex items-center justify-center mb-2 shadow-lg"
        >
          <span
            class="text-white text-3xl font-bold"
          >
            <img
              alt="Dashboard"
              class="w-10 h-10"
              src="data:image/svg+xml,%3c?xml%20version='1.0'%20encoding='utf-8'?%3e%3c!--%20Uploaded%20to:%20SVG%20Repo,%20www.svgrepo.com,%20Generator:%20SVG%20Repo%20Mixer%20Tools%20--%3e%3csvg%20fill='%23ffffff'%20width='800px'%20height='800px'%20viewBox='0%200%2032%2032'%20id='icon'%20xmlns='http://www.w3.org/2000/svg'%3e%3cdefs%3e%3cstyle%3e%20.cls-1%20{%20fill:%20none;%20}%20%3c/style%3e%3c/defs%3e%3cpolygon%20points='4%2020%204%2022%208.586%2022%202%2028.586%203.414%2030%2010%2023.414%2010%2028%2012%2028%2012%2020%204%2020'/%3e%3crect%20x='24.0001'%20y='21'%20width='2'%20height='5'/%3e%3crect%20x='20.0001'%20y='16'%20width='2'%20height='10'/%3e%3crect%20x='16'%20y='18'%20width='2'%20height='8'/%3e%3cpath%20d='M28,2H4A2.002,2.002,0,0,0,2,4V16H4V13H28.001l.001,15H16v2H28a2.0027,2.0027,0,0,0,2-2V4A2.0023,2.0023,0,0,0,28,2ZM12,11H4V4h8Zm2,0V4H28l.0007,7Z'/%3e%3crect%20id='_Transparent_Rectangle_'%20data-name='&lt;Transparent%20Rectangle&gt;'%20class='cls-1'%20width='32'%20height='32'/%3e%3c/svg%3e"
            />
          </span>
        </div>
        <h1
          class="text-3xl font-extrabold text-primary tracking-tight mb-1"
        >
          Dashboard
        </h1>
        <p
          class="text-gray-400 text-sm"
        >
          Sign in to your account
        </p>
      </div>
      <form
        class="w-full space-y-6"
      >
        <h2
          class="page-heading text-center"
        >
          Welcome Back
        </h2>
        <div
          class="mb-4"
        >
          <label
            class="block text-gray-700 text-xs sm:text-sm font-bold mb-2"
            for="email"
          >
            Email Address
          </label>
          <input
            class="
          shadow appearance-none border rounded-lg w-full py-2 sm:py-3 px-3 sm:px-4 text-xs sm:text-sm text-gray-700 leading-tight 
          focus:outline-none focus:ring-2 focus:ring-primary focus:border-transparent transition-all
          border-gray-200
          
          
        "
            id="email"
            name="email"
            placeholder="Enter your email"
            type="email"
            value=""
          />
        </div>
        <div
          class="mb-4"
        >
          <label
            class="block text-gray-700 text-xs sm:text-sm font-bold mb-2"
            for="password"
          >
            Password
          </label>
          <input
            class="
          shadow appearance-none border rounded-lg w-full py-2 sm:py-3 px-3 sm:px-4 text-xs sm:text-sm text-gray-700 leading-tight 
          focus:outline-none focus:ring-2 focus:ring-primary focus:border-transparent transition-all
          border-gray-200
          
          
        "
            id="password"
            name="password"
            placeholder="Enter your password"
            type="password"
            value=""
          />
        </div>
        <button
          class="font-bold rounded-lg focus:outline-none focus:ring-2 focus:ring-offset-2 transition-all duration-200 disabled:opacity-50 disabled:cursor-not-allowed flex items-center justify-center gap-2 whitespace-nowrap bg-primary text-white border-2 border-transparent hover:bg-white hover:text-primary hover:border-primary focus:ring-primary active:scale-95 px-4 sm:px-6 py-2 sm:py-2 text-xs sm:text-base md:text-lg w-full"
          type="submit"
        >
          Sign In
        </button>
      </form>
      <div
        class="mt-6 text-center"
      >
        <p
          class="text-gray-600 text-sm"
        >
          Don't have an account?
           
          <a
            class="text-primary hover:text-primary-dark font-semibold hover:underline transition-colors"
            data-discover="true"
            href="/signup"
          >
            Sign up here
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
